vnext/PropertySheets/PoisonUnpatchedHeaders.targets

```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- 
+ Copyright (c) Microsoft Corporation. All rights reserved.
+ Licensed under the MIT License.. 
+-->
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+
+  <!--
+    Projects trying to include headers from ReactNativeDir will fail unless
+    source patching is enabled. This is by-design to avoid conflicts with
+    patched and unpatched headers, but leads to an unintuitive error message.
+    Ensure we give a useful message when building by dyanmically generating
+    poisoned headers when patching is disabled.
+  -->
+  <Target Name="PoisonUnpatchedHeaders" Condition="'$(PatchTargetIncluded)' != 'true'" BeforeTargets="PrepareForBuild">
+    <ItemGroup>
+      <ReactNativeHeaders Include="$(ReactNativePackageDir)\ReactCommon\**\*.h" />
+      <PoisonedHeaders Include="@(ReactNativeHeaders->'$(ReactNativeDir)\ReactCommon\%(RecursiveDir)%(Filename)%(Extension)')" />
+    </ItemGroup>
+
+    <PropertyGroup>
+      <PoisonedHeaderContent><![CDATA[
+#error Facebook React Native headers may not be included without patches. \
+Ensure no React Native headers are directly or indirectly included, or import \
+"ReactPatches.targets" in your project.
+      ]]></PoisonedHeaderContent>
+    </PropertyGroup>
+
+    <MakeDir Directories="@(PoisonedHeaders->'%(RootDir)\%(Directory)')" />
+    <WriteLinesToFile
+      File="%(PoisonedHeaders.FullPath)"
+      Lines='$(PoisonedHeaderContent)'
+      Overwrite="true"
+      WriteOnlyWhenDifferent="true"/>
+  </Target>
+
```

Comment:
If you can detect the undesired condition, instead of writing poisoned headers, you could just do an <Error Message="foobar"/> instead.

Reply:
This was originally with the idea that something could include React Native property sheets and not want source patching. I didn't realize until after this PR that we don'y expose the property sheets publicly at all for new project templates using Microsoft.ReactNative.

I think I'm going to actually try to go back to just enabling patching sources for anything that includes vnext React Property sheets, since that shouldn't leak very far.